### PR TITLE
Feat: Download JSON button added

### PR DESCRIPTION
**Pull request by DhairyaMajmudar**
-> Download JSON button is now functional 
-> cloese #17
-> related to #16

cc: @seveibar

### CODE REVIEW

**Comment by seveibar** [comment]
can we remove this button? as per the discussion here: https://github.com/tscircuit/snippets/issues/16#issuecomment-2395273258

You could easily adapt this PR to make the Download Circuit JSON button though!

File: src/components/DownloadButtonAndMenu.tsx [hunk 1]
```diff
@@ -7,8 +7,18 @@ import {
...
 import { Download, ChevronDown } from "lucide-react"
+import { DownloadTSX } from "@/lib/utils/download-fn/download-tsx"
 
 export function DownloadButtonAndMenu({ className }: { className?: string }) {
+  const createBlobURL = (content: string) => {
+    const blob = new Blob([content], { type: "text/plain" })
+    return URL.createObjectURL(blob)
+  }
+
+  const { TSXContent, TSXFileName } = DownloadTSX()
```
Comment (seveibar) [inline src/components/DownloadButtonAndMenu.tsx:18]: hooks should always have a `use*` prefix and be `camelCased`, likewise for variables returned

**Comment by DhairyaMajmudar** [comment]
> can we remove this button? as per the discussion here: [#16 (comment)](https://github.com/tscircuit/snippets/issues/16#issuecomment-2395273258)
> 
> You could easily adapt this PR to make the Download Circuit JSON button though!

I got you let me modify this PR for downloading JSON data

File: src/components/DownloadButtonAndMenu.tsx [hunk 2]
```diff
@@ -7,8 +6,18 @@ import {
...
 import { Download, ChevronDown } from "lucide-react"
+import { useDownloadJSON } from "@/hooks/use-download-json"
 
 export function DownloadButtonAndMenu({ className }: { className?: string }) {
+  const createBlobURL = (content: string) => {
+    const blob = new Blob([content], { type: "text/plain" })
+    return URL.createObjectURL(blob)
+  }
+
+  const { jsonContent, jsonFileName } = useDownloadJSON()
```
Comment (seveibar) [inline src/components/DownloadButtonAndMenu.tsx:17]: Bad hook name, this hook should be named to indicate its used to download the current circuit json

useDownloadCurrentSnippetCircuitJson()

File: src/components/DownloadButtonAndMenu.tsx [hunk 3]
```diff
@@ -7,8 +6,18 @@ import {
 } from "@/components/ui/dropdown-menu"
 import { Button } from "@/components/ui/button"
 import { Download, ChevronDown } from "lucide-react"
+import { useDownloadJSON } from "@/hooks/use-download-json"
 
 export function DownloadButtonAndMenu({ className }: { className?: string }) {
+  const createBlobURL = (content: string) => {
```
Comment (seveibar) [inline src/components/DownloadButtonAndMenu.tsx:12]: Generic methods like this shouldnt be put in the inner scope of a function, move to utils directory

File: src/hooks/use-download-json.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  const snippetId = useCurrentSnippetId()
+  const { data: snippet, isLoading } = useSnippet(snippetId)
+
+  const defaultCode = useMemo(() => {
+    return decodeUrlHashToText(window.location.toString()) ?? snippet?.code
+  }, [])
+
+  const [code, setCode] = useState(defaultCode ?? "")
+
+  const { message, circuitJson } = useRunTsx(code, snippet?.snippet_type)
```
Comment (seveibar) [inline src/hooks/use-download-json.ts:17]: Very expensive recompute

**Review by seveibar — CHANGES_REQUESTED** [review]
When the download button is clicked, compute the url and direct the user to download. Recomputing and rerunning typescript is too expensive

you can probably pass in the circuit json from the parent as a prop

File: src/hooks/use-download-json.ts (hunk 4 above)
Comment (DhairyaMajmudar) [reply src/hooks/use-download-json.ts:17]: I've installed the pacakage `file-saver` which will install the files

Comment (seveibar) [reply src/hooks/use-download-json.ts:17]: the issue here is your running the tsx every time the page renders just to display a download button

Comment (seveibar) [reply src/hooks/use-download-json.ts:17]: i think the easiest solution is to have the `circuitJson` passed into the `DownloadButtonAndMenu` and remove this hook

File: src/components/DownloadButtonAndMenu.tsx [hunk 5]
```diff
@@ -7,8 +6,18 @@ import {
...
+import { createBlobURL } from "@/lib/createBlobURL"
+import create from "fake-snippets-api/routes/api/snippets/create"
 
 export function DownloadButtonAndMenu({ className }: { className?: string }) {
+  const { jsonContent, jsonFileName } = useDownloadCurrentSnippetCircuitJson()
+
+  const handleDownloadClick = () => {
+    const blob = createBlobURL(jsonContent)
+    saveAs(blob, jsonFileName)
+  }
```
Comment (seveibar) [inline src/components/DownloadButtonAndMenu.tsx:20]:
```suggestion
export function DownloadButtonAndMenu({ className, circuitJson }: { className?: string, circuitJson?: AnyCircuitElement[] }) {

  const handleDownloadClick = () => {
    const blob = createBlobURL(JSON.stringify(circuitJson))
    saveAs(blob, jsonFileName)
  }
```

**Review by seveibar — COMMENTED** [review]
see comments, we don't want to run the tsx (expensive operation) every time the download button renders

File: src/hooks/use-download-json.ts (hunk 4 above)
Comment (DhairyaMajmudar) [reply src/hooks/use-download-json.ts:17]: Cool! Thanks for the help let's do it this way.

**Comment by seveibar** [comment]
This was completed in https://github.com/tscircuit/snippets/pull/49

**Pull request closed without merging**